codec-http/src/main/java/io/netty/handler/codec/http/ReadOnlyHttpHeaders.java

```diff
@@ -0,0 +1,322 @@
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.handler.codec.CharSequenceValueConverter;
+import io.netty.util.AsciiString;
+import io.netty.util.internal.UnstableApi;
+
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import static io.netty.handler.codec.http.DefaultHttpHeaders.HttpNameValidator;
+
+@UnstableApi
+public final class ReadOnlyHttpHeaders extends HttpHeaders {
+    private final AsciiString[] keyValuePairs;
+
+    public ReadOnlyHttpHeaders(boolean validateHeaders, AsciiString... keyValuePairs) {
+        if ((keyValuePairs.length & 1) != 0) {
+            throw newInvalidArraySizeException();
+        }
+        if (validateHeaders) {
+            validateHeaders(keyValuePairs);
+        }
+        this.keyValuePairs = keyValuePairs;
+    }
+
+    private static IllegalArgumentException newInvalidArraySizeException() {
+        return new IllegalArgumentException("pseudoHeaders and otherHeaders must be arrays of [name, value] pairs");
+    }
+
+    private static void validateHeaders(AsciiString... keyValuePairs) {
+        for (int i = 0; i < keyValuePairs.length; i += 2) {
+            HttpNameValidator.validateName(keyValuePairs[i]);
+        }
+    }
+
+    private AsciiString get0(CharSequence name) {
+        final int nameHash = AsciiString.hashCode(name);
+        for (int i = 0; i < keyValuePairs.length; i += 2) {
+            AsciiString roName = keyValuePairs[i];
+            if (roName.hashCode() == nameHash && roName.contentEqualsIgnoreCase(name)) {
+                return keyValuePairs[i + 1];
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public String get(String name) {
+        AsciiString value = get0(name);
+        return value == null ? null : value.toString();
+    }
+
+    @Override
+    public Integer getInt(CharSequence name) {
+        AsciiString value = get0(name);
+        return value == null ? null : value.parseInt();
+    }
+
+    @Override
+    public int getInt(CharSequence name, int defaultValue) {
+        AsciiString value = get0(name);
+        return value == null ? defaultValue : value.parseInt();
+    }
+
+    @Override
+    public Short getShort(CharSequence name) {
+        AsciiString value = get0(name);
+        return value == null ? null : value.parseShort();
+    }
+
+    @Override
+    public short getShort(CharSequence name, short defaultValue) {
+        AsciiString value = get0(name);
+        return value == null ? defaultValue : value.parseShort();
+    }
+
+    @Override
+    public Long getTimeMillis(CharSequence name) {
+        AsciiString value = get0(name);
+        return value == null ? null : CharSequenceValueConverter.INSTANCE.convertToTimeMillis(value);
```

Comment:
just delegate to `getTimeMillis(name, null)`?

Reply:
Can you clarify? There is `getTimeMillis(CharSequence name, long defaultValue)` ... but this takes a primitive `long` (not a `Long`) so I can't pass `null`.